lib/Echidna/RPC.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Echidna.RPC where
+
+import Prelude hiding (Word)
+
+import Control.Exception (Exception)
+import Control.Lens
+import Control.Monad (foldM)
+import Control.Monad.Catch (MonadThrow, throwM)
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Reader.Class (MonadReader(..))
+import Control.Monad.State.Strict (MonadState, execStateT, runStateT, get, put, runState)
+import Data.Aeson (FromJSON(..), (.:), withObject, eitherDecodeFileStrict)
+import Data.Binary.Get (runGetOrFail)
+import Data.ByteString.Char8 (ByteString, empty)
+import Data.Has (Has(..))
+import Data.List (partition)
+import Data.Map (fromList)
+import Data.Text.Encoding (encodeUtf8)
+import EVM
+import EVM.ABI (AbiType(..), getAbi)
+import EVM.Concrete (w256)
+import EVM.Exec (exec, vmForEthrunCreation)
+import EVM.Types (Addr, W256)
+
+import qualified Control.Monad.Fail as M (MonadFail(..))
+import qualified Control.Monad.State.Strict as S (state)
+import qualified Data.ByteString.Base16 as BS16 (decode)
+import qualified Data.Text as T (Text, drop)
+import qualified Data.Vector as V (fromList)
+
+import Echidna.Exec
+import Echidna.Transaction
+
+-- | During initialization we can either call a function or create an account or contract
+data Etheno = AccountCreated Addr
+            | ContractCreated Addr Addr Integer Integer ByteString W256
+            | FunctionCall Addr Addr Integer Integer ByteString W256
+  deriving (Eq, Show)
+
+instance FromJSON Etheno where
+  -- parseJSON = genericParseJSON $ defaultOptions {fieldLabelModifier = tail, omitNothingFields = True}
+  parseJSON = withObject "Etheno" $ \v -> do
+    (ev :: String) <- v .: "event"
+    case ev of
+         "AccountCreated"  -> AccountCreated  <$> v .: "address"
+         "ContractCreated" -> ContractCreated <$> v .: "from"
+                                              <*> v .: "contract_address"
+                                              <*> (read <$> (v .: "gas_used"))
```

Comment:
That's easy to fix. It is ok to fail with an `error` if the parsing does not find the expected keywords, right or you want something more elaborated?

Reply:
`fromMaybe 0 . readMaybe` perhaps?

Reply:
So I think we do want to warn the user if they provide a value we know we can't use (that is, if `gas_used` is specified, but we can't parse it). I think using `fail` is preferable to `error`, but either is better than `fromMaybe`

Reply:
alternatively since this is the setup phase we could just scrap those values and always put in 0 for gas price and give hevm a lot of gas. we don't check if hevm's reported gas usage matches `gas_used`, this is simply parsing the JSON verbatim.

Reply:
no longer fails silently